kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/SFTPClient.scala

```diff
@@ -0,0 +1,249 @@
+        lastReplyCode = 500
+    }
+    maybeJschSession.isDefined && maybeChannelSftp.isDefined
+  }
+
+  /**
+    * Not used in this implementation of [JsCh]
+    */
+  override def setFileType(fileType: Int): Boolean = {
+    true
+  }
+
+  /**
+    * Connect to SFTP server to obtain files information (name, size, last modify)
+    * and it returns a Array[FTPFile] with all directory file info.
+    */
+  override def listFiles(pathname: String): Array[FTPFile] = {
+    maybeChannelSftp.fold {
+      logger.error(s"SFTPClient Error no channel ready to obtain files from pathname $pathname.")
+      Array[FTPFile]()
+    } { channel =>
+      if (!channel.isConnected) connectChannel(channel)
+      logger.debug(s"SFTPClient obtaining remote files from $pathname")
+      val ftpFiles = getFTPFiles(pathname, channel)
+      logger.debug(s"SFTPClient ${ftpFiles.size} remote files obtained from $pathname")
+      ftpFiles.toArray
+    }
+  }
+
+  /**
+    * Using the remote path of the file, and using [get] operator we're able to download the file and write
+    * the content into the [OutputStream].
+    */
+  override def retrieveFile(remote: String, fileBody: OutputStream): Boolean = {
+    maybeChannelSftp.fold {
+      logger.debug(s"SFTPClient Error, channel not initiated in path $remote.")
+      false
+    } { channel =>
+      channel.get(remote, fileBody)
+      logger.debug(s"SFTPClient Successful retrieving files in path $remote.")
+      true
+    }
+  }
+
+  private def getFTPFiles(pathname: String, channel: ChannelSftp): List[FTPFile] = {
+    (for {
+      _ <- Try(channel.cd(pathname))
+      ftpFiles <- fetchFiles(pathname, channel)
+    } yield ftpFiles)
+      .recover {
+        case e: Exception =>
+          logger.error(s"SFTPClient Error obtaining resources from pathname $pathname. Caused by ${ExceptionUtils.getStackTrace(e)}")
+          List()
+      }.get
+  }
+
+  private def connectChannel(channel: ChannelSftp): Unit = {
+    maybeDataTimeout
+      .fold(channel.connect())(dataTimeout => channel.connect(dataTimeout))
+  }
+
+  private def fetchFiles(pathname: String, channel: ChannelSftp): Try[List[FTPFile]] = {
+    Try {
+      channel.ls(pathname)
+        .asScala
+        .toList
+        .map(file => file.asInstanceOf[ChannelSftp#LsEntry])
+        .filter(lsEntry => lsEntry.getFilename != "." && lsEntry.getFilename != "..")
+        .map(lsEntry => createFtpFile(lsEntry))
+    }
+  }
+
+  private def createFtpFile(lsEntry: ChannelSftp#LsEntry) = {
+    val ftpFile: FTPFile = new FTPFile()
+    ftpFile.setType(0)
+    ftpFile.setName(lsEntry.getFilename)
+    ftpFile.setSize(lsEntry.getAttrs.getSize)
+
+    val dateFormat = new SimpleDateFormat("EEE MMM d HH:mm:ss zzz uuuu")
+    val calendar = Calendar.getInstance()
+    calendar.setTime(dateFormat.parse(lsEntry.getAttrs.getMtimeString))
+    ftpFile.setTimestamp(calendar)
```

Comment:
What do you think of this instead of the singleton calendar?  It's not much difference in terms of code but I prefer the Java 8 date time API to the legacy calendar.  Of course, we still go back to Calendar in the end!

```suggestion
    val dateFormat = DateTimeFormatter.ofPattern("EEE MMM d HH:mm:ss zzz uuuu")
    val date = ZonedDateTime.parse(lsEntry.getAttrs.getMtimeString, dateFormat)
    ftpFile.setTimestamp(GregorianCalendar.from(date))
```

Reply:
fine by me